lib/model/recent_senders.dart

```diff
@@ -0,0 +1,134 @@
+import 'package:collection/collection.dart';
+import 'package:flutter/foundation.dart';
+
+import '../api/model/model.dart';
+
+class MessageIdTracker {
```

Comment:
I kind of wonder if there's something off-the-shelf for what this is doing. No need to block merging on it, though, unless you or Greg see something that gives much better performance.

(I looked at [`HeapPriorityQueue`](https://api.flutter.dev/flutter/package-collection_collection/HeapPriorityQueue-class.html) but it doesn't have an efficient way to prevent duplicate entries.)

Reply:
I think this data structure is fine but there's a couple of things the algorithms should do to be more efficient.

Basically the two kinds of updates we regularly make here are:
* Adding one message (from a `MessageEvent`), which is at least nearly the latest and usually the very latest. (Might be only nearly latest because fetching messages can race with getting events.)
* Adding a whole bunch of messages (from a message fetch).

So we should make those efficient. That means:
* When the new message is the very latest, make that take $O(1)$ time instead of $\Theta(\log n)$.
* When a message fetch gets $n$ messages and there are already $m$ messages, make that not be quadratic — make it take $O((n + m) \log (n + m))$ time, instead of $\Theta((n + m)^2)$. (In fact you can make it $O(n \log n + m)$ time so long as the newly-arrived messages are all older than all the existing messages in the given tracker, which should be the common case here.)

  The key there is, first, thread things through so the `RecentSenders` gets a single method call with the whole batch of messages, rather than one per message. (There'll still be a `handleMessageEvent` for the `MessageEvent` case.) Then, similarly, for each `MessageIdTracker` that's affected, make a single method call to that `MessageIdTracker` with all the relevant messages.